docs/agents/agent-development.md

```diff
@@ -237,14 +237,37 @@ Global labels can be specified via the environment variable `ELASTIC_APM_GLOBAL_
 
 ### Transactions
 
+#### Transaction outcome
+
+The optional `outcome` property denotes whether the transaction represents a success or a failure.
+The APM Server converts this to the [`event.outcome`](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-outcome.html) field.
+
+- `"failure"`: Indicates that this transaction describes a failed result.
+- `"success"`: Indicates that this transaction describes a successful result.
+- `"unknown"`:
+  TODO: does unknown make sense in the context of spans and transactions?
+  Would `unknown` apply to an HTTP 202 Accepted? Or a client span that puts a message into a queue?
+  In these cases we only know whether a message could successfully be enqueued but not if it could successfully be processed.
+  Or is this reserved to events that only represent the request part and where the response is a separate event?
+  A transaction contains information about both the request and the response.
+  If `unknown` doesn't apply to transactions it might make more sense to use a boolean as the datatype of this field in the intake API and let APM Server convert to `failure`/`success`. 
+- `null`/`undefined`: There's no information about the outcome of the transaction. \
+  Note that this is not equal to [`event.outcome: unknown`](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-outcome.html#ecs-event-outcome-unknown)
+  as it would imply that the event is not a `"failure"`.
+  Not setting the `outcome`, or setting it to `null`, makes no statement on the outcome of the transaction.
+  This may be the case when tracking a custom transaction without explicitly setting an outcome.
+
+What counts as a failed or successful request depends on the protocol and does not depend on whether there are error documents associated with a transaction.
+
 #### HTTP Transactions
 
 Agents should instrument HTTP request routers/handlers, starting a new transaction for each incoming HTTP request. When the request ends, the transaction should be ended, recording its duration.
 
 - The transaction `type` should be `request`.
 - The transaction `result` should be `HTTP Nxx`, where N is the first digit of the status code (e.g. `HTTP 4xx` for a 404)
+- The transaction `outcome` should be `"success"` for HTTP status codes < 500 and `"failure"` for status codes >= 500. \
+  Status codes in the 4xx range (client errors) are not considered a `failure` as the failure has not been caused by the application itself but by the caller.
```

Comment:
I think I agree with this decision, but I worry that this will be confusing to users. I think the first instinct is to assume a 4xx is a failure. How will we communicate this to users?

Reply:
Good point and yes, we should!

Reply:
Done, please resolve if you think it's clear enough now.

Reply:
Since `event.outcome` is extremely coarse grained with its 3 acceptable values, such grey areas are bound to happen. We just need to do our best to map the protocol or event source to these 3 values. The other event source or protocol detail fields will capture all the necessary shades of failure/success in detail.

So I agree with this 4xx being mapped to success. 👍